lib/plug/router.ex

```diff
@@ -601,14 +624,45 @@ defmodule Plug.Router do
         conn = update_in(unquote(conn).params, merge_params)
         conn = update_in(conn.path_params, merge_params)
 
-        Plug.Router.__put_route__(conn, unquote(path), fn var!(conn), var!(opts) ->
-          _ = var!(opts)
-          unquote(body)
-        end)
+        Plug.Router.__put_route__(conn, unquote(path), unquote(callback))
+      end
+    end
+  end
+
+  @doc false
+  def __to__(module, options) do
+    {to, options} = Keyword.pop(options, :to)
+    {init_opts, options} = Keyword.pop(options, :init_opts, [])
+
+    router_to = Module.get_attribute(module, :plug_router_to)
+    callback = :"plug_router_to_#{map_size(router_to)}"
+    router_to = Map.put(router_to, callback, {to, init_opts})
+    Module.put_attribute(module, :plug_router_to, router_to)
+    {Macro.var(callback, nil), options}
+  end
+
+  defp wrap_function_do(body) do
+    quote do
+      fn var!(conn), var!(opts) ->
+        _ = var!(opts)
+        unquote(body)
       end
     end
   end
 
+  defp expand_options(opts, caller) do
+    if Macro.quoted_literal?(opts) do
+      Macro.prewalk(opts, &expand_alias(&1, caller))
+    else
+      opts
+    end
+  end
```

Comment:
After playing around a bit more I realised it only works without using function calls in assigns.

Doing something like this
 
```elixir
match("/some/route", to: SomePlug, assigns: %{key: SomeModule.value()})
```

creates a compile time dependency on **both** `SomePlug` and `SomeModule` since the function call isn't a literal and therefore `opts` as a whole isn't a literal.

I'm not sure if it's possible or desirable to remove the compile time dependency on both but doing something like this means it only has a compile time dependency on `SomeModule` and not `SomePlug`.

```suggestion
  defp expand_options(opts, caller) do
    Enum.map(opts, fn {k, v} -> {k, do_expand_options(v, caller)} end)
  end

  defp do_expand_options(opts, caller) do
    if Macro.quoted_literal?(opts) do
      Macro.prewalk(opts, &expand_alias(&1, caller))
    else
      opts
    end
  end

```

Reply:
Exactly. If you are calling a module, then you do have a compile-time dependency and removing it would lead to bugs. In your case, I would recommend moving the setting of the assigns to a `plug`, especially if you can do so based on a certain rule.

Reply:
@hoyon however, it is working without the assigns, right? If so let me know so I can merge this! :D

Reply:
@josevalim Yep without the assigns it's working!